spatialprofilingtoolbox/cggnn/extract.py

```diff
@@ -0,0 +1,57 @@
+"""Extract information cg-gnn needs from SPT."""
+
+from typing import Any
+
+from pandas import DataFrame, concat, merge  # type: ignore
+from numpy import sort  # type: ignore
+
+from spatialprofilingtoolbox.db.feature_matrix_extractor import FeatureMatrixExtractor
+
+
+def _create_cell_df(
+    dfs_by_specimen: dict[str, DataFrame],
+    feature_names: dict[str, str],
+) -> DataFrame:
+    """Find simple and complex phenotypes, and locations and merge into a DataFrame."""
+    feature_ids_to_names = {ft_id: 'FT_' + ft_name for ft_id, ft_name in feature_names.items()}
+    for specimen, df_specimen in dfs_by_specimen.items():
+        df_specimen.rename(feature_ids_to_names, axis=1, inplace=True)
+        # TODO: Create complex phenotype columns
+        df_specimen.rename({'pixel x': 'center_x', 'pixel y': 'center_y'},
+                           axis=1, inplace=True)
+        df_specimen['specimen'] = specimen
+
+    # TODO: Reorder so that it's simple phenotype, specimen, complex phenotype, x, y
+    df = concat(dfs_by_specimen.values(), axis=0)
+    df.index.name = 'histological_structure'
+    return df
+
+
+def _create_label_df(
+    df_assignments: DataFrame,
+    df_strata: DataFrame,
+) -> tuple[DataFrame, dict[int, str]]:
+    """Get slide-level results."""
+    df = merge(df_assignments, df_strata, on='stratum identifier', how='left')[
+        ['specimen', 'subject diagnosed result']].rename(
+        {'specimen': 'slide', 'subject diagnosed result': 'result'}, axis=1)
```

Comment:
(This code is from what was already implemented in `main` in `run_sql.py`, just moved to a new file.) Is this how we want to define the labels to classify against? It appears to be a little different from what we discussed over email.

Reply:
@jimmymathews Regarding stratification, this implementation uses only "subject diagnosed result" but not "stratum_identifier", "local temporal position indicator", or "subject diagnosed condition".

What do we want the model `spt cggnn extract` supports to do? In prior iterations, it's been to predict, on pre-treatment slides only, if the patient the tissue sample is from will respond to treatment. Do we want to continue this, or generalize?

Reply:
Prediction of treatment response from pre-treatment slides is certainly not the only study design that we support.

Although one cannot, in advance, write a single description of the outcome variable that will always be of greatest interest in any study, the `SampleStratificationCreator` does a pretty good job of automatically creating a best guess at a reasonable granularity, given the facts about the samples available according to the schema.

The result of `SampleStratificationCreator`'s work is the outcome variable that I endorse for the use case of general-purpose downstream analysis.

It is defined by the fields named `stratum_identifier` (an integer) and `sample`, in a supplementary table called `sample_strata`.

In this use case, one shouldn't be reading `local_temporal_position_indicator` or `subject_diagnosed_condition` or `subject_diagnosed_result` at all. In this use case, these values are just metadata tags for the strata.

Reply:
(`sample` here means the tissue sample, right? I don't see why the `sample` is relevant to the stratification aside from samples being the granularity at which we're stratifying.)

As I envision it, the `spt cggnn extract` command isn't for "general-purpose downstream analysis" but to extract the specific information the user wants to train a model on. (If this isn't the case, we should write a separate command that does so, as I don't think this should be within the scope of the standalone `cg-gnn` package.)

So, if prediction of treatment response isn't the only study design we want to support, it sounds like what we should do is provide the person calling `spt cggnn extract` some way to identify which strata they want to use. For example, in the treatment response use case, they should be given the option of only selecting strata from before treatment. In most use cases, I don't think the user would want every stratum possible.

Reply:
I see, I misunderstood your earlier remark that you want "to not have to study the intricacies of the database to run the cg-gnn pipeline". I understood it to mean that you want labels like `1`, `2`, and `3`, and you don't want to be involved with coming up with finer outcome data using the actual study data.

But it seems that you meant that you want *at runtime of this extract tool* to not have to study the intricacies of the database, but that you are willing to study them when writing the extract tool source code.

In this case, the convenience sample stratifier isn't for you. Just use direct SQL queries to `diagnosis`, `intervention`, `subject`, `specimen_collection_process`, etc. You could probably make a nice little CLI dialog that solicits from the user on the basis of what's already known to be available.

Reply:
The whole design constraint for `SampleStratificationCreator` was automation with no user intervention, but that's the opposite of what you're describing.

Reply:
Would this be built into the `FeatureMatrixExtractor` like the convenience stratifier is, or handled seperately?

And to do that we'd have to guess at all the different ways a user would want to slice up and stratify the samples... or just give them every option in the database. At which point, the user may as well write their own SQL queries.

Sure we can't just support only treatment response for now, and leave the implementation of alternative classifications as ~an exercise for the reader~ future work?

Reply:
I don't think any of our other pre-curated studies have 2 cohorts that are both pre-treatment and distinguished by response.

Reply:
Maybe 1 other?

Reply:
So we'd have to think of at least one classification method for each study, and they're all different.